File: go/v6/csv_reader.go
package main

import (
	"bufio"
	"bytes"
	"os"
	"runtime"
	"strconv"
	"sync"
	"syscall"
)

type Segment struct {
	start int64
	size  int64
}

func Segments(filePath string) ([]Segment, error) {
	const oneMB = 1048576
	const scalarForWorkers = 1

	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	info, err := file.Stat()
	if err != nil {
		return nil, err
	}
	totalSize := info.Size()
	cores := runtime.NumCPU() * scalarForWorkers

	targetSize := totalSize / int64(cores)
	if targetSize < oneMB {
		targetSize = oneMB
	}

	var segments []Segment
	var currentPos int64 = 0

	for currentPos < totalSize {
		start := currentPos
		endCandidate := currentPos + targetSize

		if endCandidate >= totalSize {
			segments = append(segments, Segment{start: start, size: totalSize - start})
			break
		}

		_, err := file.Seek(endCandidate, 0)
		if err != nil {
			return nil, err
		}

		buffer := make([]byte, 1)
		extraBytes := int64(0)
		for {
			_, err := file.Read(buffer)
			if err != nil {
				break
			}
			extraBytes++
			if buffer[0] == '\n' {
				break
			}
		}

		actualEnd := endCandidate + extraBytes
		segments = append(segments, Segment{start: start, size: actualEnd - start})
		currentPos = actualEnd 
	}

	return segments, nil
}

func LoadPoints(path string) ([]Point, error) {
    segments, err := Segments(path)
    if err != nil {
        return nil, err
    }

	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	info, err := file.Stat()
	if err != nil {
		return nil, err
	}
	totalSize := info.Size()

	data, err := syscall.Mmap(int(file.Fd()), 0, int(totalSize), syscall.PROT_READ, syscall.MAP_SHARED)
	if err != nil {
		return nil, err
	}
	defer syscall.Munmap(data)

    numSegments := len(segments)
    var points []Point
    var mu sync.Mutex
    var wg sync.WaitGroup

    for i := 0; i < numSegments; i++ {
        wg.Add(1)
        go func(segment Segment) {
            defer wg.Done()

            workerData := data[segment.start : segment.start+segment.size]
			reader := bytes.NewReader(workerData)
            scanner := bufio.NewScanner(reader)

            var pointsPartition []Point
            for scanner.Scan() {
                line := bytes.TrimSpace(scanner.Bytes())
                if len(line) > 0 {
                    p, err := processLine(line)
                    if err != nil {
                        continue
                    }
                    if p != nil {
                        pointsPartition = append(pointsPartition, *p)
                    }
                }
            }
            mu.Lock()
            points = append(points, pointsPartition...)
            mu.Unlock()

        }(segments[i])
    }

    wg.Wait()

    return points, nil
}

func processLine(line []byte) (*Point, error) {
	parts := bytes.Split(line, []byte(","))
	coords := make([]float64, len(parts))

	for i, part := range parts {
		val, err := strconv.ParseFloat(string(bytes.TrimSpace(part)), 64)
		if err != nil {
			return nil, err
		}
		coords[i] = val
	}

	point, err := NewPoint(coords)
	if err != nil {
		return nil, err
	}

	return point, nil
}

File: java/v7/src/main/java/com/kmeans/CSVReader.java
package com.kmeans;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;

public class CSVReader {

    private static final int SCALAR_FOR_WORKERS_READ = 1;
    private static final int ONE_MB = 1048576;
    private record Segment(long start, long size) {}

    private static List<Segment> getSegments(Path path) throws IOException {
        List<Segment> segments = new ArrayList<>();
        try (RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r")) {
            long totalSize = raf.length();
            int cores = Runtime.getRuntime().availableProcessors() * SCALAR_FOR_WORKERS_READ;
            long targetSize = totalSize / cores;
            if (targetSize < ONE_MB) {
                targetSize = ONE_MB;
            }
            long currentPos = 0; 

            while (currentPos < totalSize) {
                long start = currentPos;
                long endCandidate = currentPos + targetSize;

                if (endCandidate >= totalSize) {
                    segments.add(new Segment(start, totalSize - start));
                    break;
                }

                raf.seek(endCandidate);
                long extraBytes = 0;
                while (true) {
                    int b = raf.read();
                    if (b == -1) {
                        break;
                    }
                    extraBytes++;
                    if (b == '\n') {
                        break;
                    }
                }

                long actualEnd = endCandidate + extraBytes;
                segments.add(new Segment(start, actualEnd - start));
                currentPos = actualEnd;
            }
        }
        return segments;
    }

    public static List<Point> loadPoints(Path path) throws Exception {
        List<Segment> segments = getSegments(path);
        int numSegments = segments.size();
        List<Point> points = new ArrayList<>();
        Object mutex = new Object();
        
        Thread[] threads = new Thread[numSegments];

        for (int i = 0; i < numSegments; i++) {
            Segment segment = segments.get(i);
        
            threads[i] = Thread.ofPlatform().start(() -> {
                List<Point> pointsPartition = new ArrayList<>();
                try (FileChannel ch = FileChannel.open(path, StandardOpenOption.READ)) {
                    ByteBuffer buffer = ch.map(FileChannel.MapMode.READ_ONLY, segment.start(), segment.size());
                
                    if (segment.start() == 0) {
                        while (buffer.hasRemaining() && buffer.get() != '\n');
                    }
                
                    ByteArrayOutputStream lineBuffer = new ByteArrayOutputStream();
                    while (buffer.hasRemaining()) {
                        byte b = buffer.get();
                        if (b == '\n') {
                            Point p = processLine(lineBuffer.toByteArray());
                            if (p != null) {
                                pointsPartition.add(p);
                            }
                            lineBuffer.reset();
                        } else if (b != '\r') {
                            lineBuffer.write(b);
                        }
                    }
                    if (lineBuffer.size() > 0) {
                        Point p = processLine(lineBuffer.toByteArray());
                        if (p != null) {
                            pointsPartition.add(p);
                        }
                    }
                    
                    synchronized (mutex) {
                        points.addAll(pointsPartition);
                    }
                    
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        }

        for (int i = 0; i < threads.length; i++) {
            if (threads[i] == null) continue;
            while (threads[i].isAlive()) {
                Thread.yield();
            }
        }

        return points;
    }

    public static Point processLine(byte[] lineBytes) {
        String currentLine = new String(lineBytes).trim();
        if (currentLine.isEmpty()) {
            return null;
        }

        try {
            String[] parts = currentLine.split(",");
            double[] coords = new double[parts.length];
            for (int i = 0; i < parts.length; i++) {
                coords[i] = Double.parseDouble(parts[i].trim());
            }
            return new Point(coords); 
        } catch (NumberFormatException e) {
            return null;
        }
    }
}
